README.md
plugin.video.starmovies.org
---------------------------

A Kodi Addon to display and manage themoviedb lists and ratings

Usage
-----

Settings:
---------

Installation
------------

 - Download the add-on as a ZIP file from the top of this page
 - Open Kodi
 - Go to `System -> Settings -> Add-ons -> Install from zip file`
 - Restart Kodi and enjoy :)
 
Release history
---------------
  * 2016-09-01 v 1.0.1 Initial release
  6.0.0 - Kodi 19


changelog.txt
[B]6.1.0[/B] - Kodi 20
[B]6.0.0[/B] - Kodi 19
[B]1.0.3[/B] - New Site starmovies.org
[B]1.0.1[/B] - Initial Release


addon.py
# -*- coding: utf-8 -*-
import sys, os, re
import json
import time
import ssl
import xbmcplugin
import xbmcaddon
import xbmcgui
import xbmcvfs
from xbmcplugin import addDirectoryItem, endOfDirectory

try:
    from urllib.parse import parse_qs, urlencode, quote, unquote
    import urllib.request
    python="3"
except ImportError:
    from urllib import urlencode, quote, unquote
    from urlparse import parse_qs
    import urllib2
    python="2"

addon = xbmcaddon.Addon()
addon_handle = int(sys.argv[1])
args = parse_qs(sys.argv[2][1:])
menu = args.get('menu', ['None'])[0]
lang = addon.getLocalizedString

def setUrl(pQuery):
    return sys.argv[0] + '?' + urlencode(pQuery)

def getRequest2(url):
    try:
        context = ssl._create_unverified_context()
        request = urllib2.Request(url)
        response = json.load(urllib2.urlopen(request, context=context))
        return response
    except Exception as e:
        xbmc.log(str(e),3)

def getRequest3(url):
    try:
        request = urllib.request.Request(url)
        response = json.load(urllib.request.urlopen(request))
        return response
    except Exception as e:
        xbmc.log(str(e),3)
        
def list_items(listType, videoType, order):
    if videoType=="S":
        dbType="tvshow"
        xbmcplugin.setContent(addon_handle, "tvshows")
    else:
        dbType="movie"
    xbmcplugin.setContent(addon_handle, "movies")	
    url = "https://www.starmovies.org/WebService.asmx/getList?listType="+listType+"&videoType="+videoType+"&order="+order+"&pg=0&pgSize=5000&usr="+addon.getSetting('usr')+"&pwd="+addon.getSetting('pwd')
    if python=="3":
        items = getRequest3(url)
    else:
        items = getRequest2(url)
    if items==None:
        li = xbmcgui.ListItem("Check settings user or password")
        addDirectoryItem(addon_handle, "", li, isFolder=False)
    else:
        for item in items:
            if listType=="watched" and videoType=="S" and item["season"]!=None:
                title=item["title"]+" - "+"s"+str(item["episode"])+"e"+str(item["season"])+" "+str(item["episodetitle"])
            else: title=item["title"]
            infolabels={'IMDBNumber': item['imdbId'], 'title': title, 'year': item['release_date'], 'rating': item['userrating'], "mediatype": dbType}
            li = xbmcgui.ListItem(title)
            li.setInfo('video', infolabels)
            if item["poster"]!=None:
                li.setArt({ "poster" : item["poster"].strip() })
                li.setArt({ "thumbnail" : item["poster"].strip() })
                li.setArt({ "poster" : item['poster']})
            if item["backdrop"]!=None:
                li.setArt({ "fanart" : item["backdrop"].strip() })		
            li.setProperty('IsPlayable', 'true')
            link=item['link']
            if link==None: link="noLink"
            addDirectoryItem(addon_handle, quote(link), li, isFolder=False)
    endOfDirectory(addon_handle)

def Main():
    if menu == 'None':
        xbmcplugin.setContent(addon_handle, "menu")
        imgPath=xbmcvfs.translatePath(xbmcaddon.Addon().getAddonInfo('path'))
        li = xbmcgui.ListItem(lang(30004))
        li.setArt({'icon':imgPath+'/resources/movies.png'})
        addDirectoryItem(addon_handle, setUrl({'menu': 'ratedMovies'}), li, True)
        li = xbmcgui.ListItem(lang(30005))
        li.setArt({'icon':imgPath+'/resources/episodes.png'})
        addDirectoryItem(addon_handle, setUrl({'menu': 'ratedTv'}), li, True)
        li = xbmcgui.ListItem(lang(30007))
        li.setArt({'icon':imgPath+'/resources/movies.png'})
        addDirectoryItem(addon_handle, setUrl({'menu': 'watchlistMovies'}), li, True)
        li = xbmcgui.ListItem(lang(30007))
        li.setArt({'icon':imgPath+'/resources/episodes.png'})
        addDirectoryItem(addon_handle, setUrl({'menu': 'watchlistTv'}), li, True)
        li = xbmcgui.ListItem(lang(30008))
        li.setArt({'icon':imgPath+'/resources/movies.png'})
        addDirectoryItem(addon_handle, setUrl({'menu': 'favoriteMovies'}), li, True)
        li = xbmcgui.ListItem(lang(30009))
        li.setArt({'icon':imgPath+'/resources/episodes.png'})
        addDirectoryItem(addon_handle, setUrl({'menu': 'favoriteTv'}), li, True)
        li = xbmcgui.ListItem(lang(30018))
        li.setArt({'icon':imgPath+'/resources/movies.png'})
        addDirectoryItem(addon_handle, setUrl({'menu': 'watchedMovies'}), li, True)
        li = xbmcgui.ListItem(lang(30019))
        li.setArt({'icon':imgPath+'/resources/episodes.png'})
        addDirectoryItem(addon_handle, setUrl({'menu': 'watchedTv'}), li, True)
        endOfDirectory(addon_handle)
    elif menu == 'ratedMovies':
        list_items("rated","M", "usrrating")
    elif menu == 'ratedTv':
        list_items("rated","S", "usrrating")
    elif menu == 'watchlistMovies':
        list_items("watchlist","M", "date")
    elif menu == 'watchlistTv':
        list_items("watchlist","S", "date")
    elif menu == 'favoriteMovies':
        list_items("favorites","M", "date")
    elif menu == 'favoriteTv':
        list_items("favorites","S", "date")
    elif menu == 'watchedMovies':
        list_items("watched","M", "watched")
    elif menu == 'watchedTv':
        list_items("watched","S", "watched")

if addon.getSetting('usr')=='' or addon.getSetting('pwd')=='':
    xbmc.executebuiltin('Notification(' + lang(30016) + ',)')
    xbmcaddon.Addon().openSettings() 
else:
    Main()


synch.py
# -*- coding: utf-8 -*-
import sys, os, re
import json
import time
import ssl
import xbmc
import xbmcplugin
import xbmcaddon
from xbmcgui import ListItem
from xbmcplugin import addDirectoryItem, endOfDirectory

try:
    import urllib.request
    from urllib.parse import parse_qs, urlencode, quote, unquote
    python="3"
except ImportError:
    import urllib2
    from urllib import urlencode, quote, unquote
    python="2"

addon = xbmcaddon.Addon("plugin.video.starmovies.org")

def getRequest2(url):
    try:
        context = ssl._create_unverified_context()
        request = urllib2.Request(url)
        response = urllib2.urlopen(request, context=context)
        return response
    except:
        pass

def getRequest3(url):
    try:
        req = urllib.request.Request(url)
        with urllib.request.urlopen(req) as response:
            return response
    except:
        pass

def jsonrpc(query):
    return json.loads(xbmc.executeJSONRPC(json.dumps(query)))

def jsonrpc2(query):
    return xbmc.executeJSONRPC(json.dumps(query))

def setUrl(pQuery):
    return sys.argv[0] + '?' + urlencode(pQuery)

def synchCollection(videoType):
    if videoType=="M":
        query = {'jsonrpc': '2.0','id': 0,'method': 'VideoLibrary.GetMovies','params': {'properties': ['imdbnumber', 'userrating', 'playcount', 'lastplayed', 'dateadded', 'file'] } }
        movies = "[" + jsonrpc2(query) + "]"
    else:
        query = {'jsonrpc': '2.0','id': 0,'method': 'VideoLibrary.GetTvShows','params': {'properties': ['imdbnumber', 'userrating', 'playcount', 'lastplayed', 'dateadded', 'file'] } }
        movies = "[" + jsonrpc2(query) + "]"
    usr = addon.getSetting('usr')
    pwd = addon.getSetting('pwd')
    movies = movies.replace("ã","a")
    movies = movies.replace("â","a")
    movies = movies.replace("ä","a")
    movies = movies.replace("á","a")
    movies = movies.replace("à","a")
    movies = movies.replace("â","a")
    movies = movies.replace("è","e")
    movies = movies.replace("é","e")
    movies = movies.replace("ê","e")
    movies = movies.replace("í","i")
    movies = movies.replace("ó","o")
    movies = movies.replace("ò","o")
    movies = movies.replace("ô","o")
    movies = movies.replace("õ","o")
    movies = movies.replace("ô","o")
    movies = movies.replace("ù","u")
    movies = movies.replace("ú","u")
    movies = movies.replace("ç","c")
    movies = movies.replace("³","")
    movies = movies.replace("Á","A")
    movies = movies.replace("À","A")
    movies = movies.replace("Ê","E")
    movies = movies.replace("È","E")
    movies = movies.replace("É","E")
    movies = movies.replace("î","i")
    movies = movies.replace("Ô","O")
    movies = movies.replace("","")
    movies = movies.replace("–","")
    movies = movies.replace("´","'")
    movies = movies.replace("ñ","n")
    movies = movies.replace("½","")
    movies = movies.replace("ø","")
    query_args = {'videoType': videoType, 'usr': quote(usr), 'pwd': quote(pwd), 'json': movies}
    url = "https://www.starmovies.org/WebService.asmx/synchCollection"
    try:
        if python=="3":
            data = urllib.parse.urlencode(query_args).encode("utf-8")
            req = urllib.request.Request(url)
            response=urllib.request.urlopen(req, data=data)
            xbmc.log(str(response.read()), 3)
            ####xbmc.executebuiltin('Notification(Synch completed,'+response.read()+')')
        else:
            data = urlencode(query_args)
            context = ssl._create_unverified_context()
            request = urllib2.Request("https://www.starmovies.org/WebService.asmx/synchCollection", data)
            response = urllib2.urlopen(request, context=context)
            xbmc.executebuiltin('Notification(Synch completed,'+response.read()+')')
    except urllib.error.HTTPError as error:
        xbmc.executebuiltin('Notification(ERROR:,'+error.read()+')')

def synch(listType, videoType):
    if videoType=="M":
        query = {'jsonrpc': '2.0','id': 0,'method': 'VideoLibrary.GetMovies','params': {'properties': ['imdbnumber', 'userrating', 'playcount', 'lastplayed', 'dateadded', 'file'] } }
        movies = jsonrpc(query)['result'].get("movies", [])
    else:
        query = {'jsonrpc': '2.0','id': 0,'method': 'VideoLibrary.GetTvShows','params': {'properties': ['imdbnumber', 'userrating', 'playcount', 'lastplayed', 'dateadded', 'file'] } }
        movies = jsonrpc(query)['result'].get("tvshows", [])

    for movie in movies:
        if movie["imdbnumber"]!="" and (listType=="collection" or (listType=="watched" and movie["playcount"]>0) or (listType=="rated" and movie["userrating"]>0)):
            url = "https://www.starmovies.org/WebService.asmx/synchList?listType=" + listType + "&videoType=" + videoType
            url = url + "&usr=" + quote(addon.getSetting('usr'))
            url = url + "&pwd=" + addon.getSetting('pwd')
            if videoType=="M":
                url = url + "&kodiId=" + str(movie["movieid"]) 
            else:
                url = url + "&kodiId=" + str(movie["tvshowid"]) 
            url = url + "&externalId=" + movie["imdbnumber"] 
            url = url + "&rating=" + str(movie["userrating"])
            plDate=str(movie["lastplayed"])[:10]
            if plDate=="" : plDate="2000-01-01" 
            addDate=str(movie["dateadded"])[:10]
            if addDate=="": addDate="2000-01-01" 
            if (listType=="watched"): url = url + "&date=" + plDate + "&link=" + urllib.quote_plus(movie["file"])
            if (listType=="collection"): url = url + "&date=" + addDate + "&link="
            if (listType=="rated"): url = url + "&date=" + "&link="
            if python=="3":
                response = getRequest3(url)
            else:
                response = getRequest2(url)
            xbmc.log("=============================================================================",3)
            xbmc.log(url,3)
            xbmc.log(str(response),3)

if len(sys.argv)>1:
    addon = xbmcaddon.Addon("plugin.video.starmovies.org")
    xbmc.executebuiltin('Notification(Synch started,'+sys.argv[1]+')')
    if sys.argv[1]=="collection":
        synchCollection(sys.argv[2])
    else:
        synch(sys.argv[1], sys.argv[2])
    xbmc.executebuiltin('Notification(Synch completed,'+sys.argv[1]+')')


context.py
# -*- coding: utf-8 -*-
import os
import sys
import json
import xbmc
import xbmcgui
import xbmcaddon
import xbmcvfs
import urllib
import urllib2
import ssl

addon = xbmcaddon.Addon()
lang = addon.getLocalizedString
USR=addon.getSetting('usr')
PWD=addon.getSetting('pwd')

baseURL = 'https://api.themoviedb.org/3/'
basePARM = '?api_key=fb77526161109488c45abd0f75960a0f'
HEADERS = {'Accept': 'application/json','Content-Type': 'application/json'}

def getTMDBbyId(dbtype, IMDBNumber):
    tmdbId = ''
    if dbtype=='tv':
        request = urllib2.Request(baseURL + 'find/' + IMDBNumber + basePARM + '&external_source=tvdb_id')
        response = json.loads(urllib2.urlopen(request, timeout=3).read())
        tmdbId = response["tv_results"][0]["id"]
    else:
        request = urllib2.Request(baseURL + 'find/' + IMDBNumber + basePARM + '&external_source=imdb_id')
        response = json.loads(urllib2.urlopen(request, timeout=3).read())
        tmdbId = response["movie_results"][0]["id"]
    return tmdbId

def getTMDBbyName(dbType, title, year):
    data = {}
    tmdbId = ''
    if dbType=="tvshow": dbType="tv"
    if dbType!='tv':
        url = baseURL + 'search/' + dbType + basePARM + "&query=" + title.replace(" ","%20") + "&year=" + year
        request = urllib2.Request(url=url, data=json.dumps(data), headers=HEADERS)
        response = json.loads(urllib2.urlopen(request, timeout=3).read())
        if response["total_results"]!=0: tmdbId = response["results"][0]["id"]
    if tmdbId == '':
        url = baseURL + 'search/' + dbType + basePARM + "&query=" + title.replace(" ","%20")
        request = urllib2.Request(url=url, data=json.dumps(data), headers=HEADERS)
        response = json.loads(urllib2.urlopen(request, timeout=3).read())
        if response["total_results"]!=0: tmdbId = response["results"][0]["id"]
    return tmdbId

def api(func, lstType, rating, videoType, tmdbId, kodiId, imdbId):
    url = 'https://www.starmovies.org/WebService.asmx/kodiMark?func='+ func + '&lstType=' + lstType +'&videoType=' + videoType + '&tmdbId=' + str(tmdbId) + '&imdbId='+imdbId+'&tvdbId=0&kodiId='+str(kodiId)+'&rating=' + str(rating) + '&usr='+USR+'&pwd='+PWD
    context = ssl._create_unverified_context()
    request = urllib2.Request(url=url)
    response = urllib2.urlopen(request, context=context, timeout=3).read()
    xbmc.executebuiltin('Notification(' + lang(30010) + ',' + response + ')')
    
def prompt(dbid, dbType, tmdbId, title, imdbId):
    if dbType=="tvshow": videoType="S"
    else: videoType="M"
    url="https://www.starmovies.org/Webservice.asmx/getStates?videoType="+videoType+"&tmdbId="+str(tmdbId)+"&imdbId="+imdbId+"&usr="+USR+"&pwd="+PWD
    context = ssl._create_unverified_context()
    request = urllib2.Request(url)
    response = json.loads(urllib2.urlopen(request, context=context, timeout=3).read())
    if str(response["watchlist"])=='None':
        pList = [lang(30011)]
        func = "Mark"
    else:
        pList = [lang(30012)]
        func = "unMark"
    if str(response["favorites"])=='None':
        pList.append(lang(30013))
        func = "Mark"
    else:
        pList.append(lang(30014))
        func = "unMark"
    if str(response["watched"])=='None':
        pList.append(lang(30022))
        func = "Mark"
    else:
        pList.append(lang(30023))
        func = "unMark"
    if str(response["rated"])!='None': pList.append(lang(30015))
    for i in reversed(range(1, 11)):
        pList.append(str(i))
    sel = xbmcgui.Dialog().select(lang(30010)+" - "+title, pList)
    # Handle Selection
    if sel == -1: return
    elif sel == 0:
        api(func, "watchlist", 0, videoType, tmdbId, dbid, imdbId)
    elif sel == 1:
        api(func, "favorites", 0, videoType, tmdbId, dbid, imdbId)
    elif sel == 2:
        api(func, "watched", 0, videoType, tmdbId, dbid, imdbId)
    else:
        if lang(30015)==pList[sel]:
            rating = 0
        else:
            rating = int(pList[sel])
        if rating==0: func="unMark"
        else: func="Mark"
        api(func, "rated", rating, videoType, tmdbId, dbid, imdbId)

# Prepares and shows the selection list
def main():
    dbid=xbmc.getInfoLabel('ListItem.DBID')
    dbType = xbmc.getInfoLabel('ListItem.DBTYPE')
    if dbType!="tvshow": IMDBNumber = xbmc.getInfoLabel('ListItem.IMDBNumber')
    else: IMDBNumber = ""
    tit = xbmc.getInfoLabel('ListItem.Title')
    if (tit==""): tit = xbmc.getInfoLabel('ListItem.Label')

    tmdbId = xbmc.getInfoLabel('ListItem.Top250');
    if str(tmdbId) == '':
        if str(IMDBNumber) == '':
            tmdbId = getTMDBbyName(dbType, tit, xbmc.getInfoLabel('ListItem.Year'))
        else:
            tmdbId = getTMDBbyId(dbType, IMDBNumber)

    if str(tmdbId) == '':
        xbmc.executebuiltin('Notification(' + lang(30010) + ',' + lang(30017) + ')') # Title not found
    else:
        prompt(dbid, dbType, tmdbId, tit.decode("utf-8"), IMDBNumber)

if __name__ == '__main__':
    if USR == '' or PWD == '' :
        xbmc.executebuiltin('Notification(' + lang(30016) + ',)') # Set user and passsword
        xbmcaddon.Addon().openSettings() 
    else:
        main()


service.py
# -*- coding: utf-8 -*-
import synch
import sys, os, re
import json
import time
import xbmcplugin
import xbmcaddon
from xbmcgui import ListItem
from xbmcplugin import addDirectoryItem, endOfDirectory

addon = xbmcaddon.Addon("plugin.video.starmovies.org")

def jsonrpc(query):
	return json.loads(xbmc.executeJSONRPC(json.dumps(query, encoding='utf-8')))

def jsonrpc2(query):
	return xbmc.executeJSONRPC(json.dumps(query, encoding='utf-8'))

def setUrl(pQuery):
	return sys.argv[0] + '?' + urllib.urlencode(pQuery)
 
class ScanMonitor(xbmc.Monitor):
	def __init__(self, *args, **kwargs):
		xbmc.Monitor.__init__(self)

	def onCleanFinished(self, database):
		synch.synchCollection("S")
		synch.synchCollection("M")

	def onScanFinished(self, database):
		synch.synchCollection("S")
		synch.synchCollection("M")

scan_monitor = ScanMonitor()

while not xbmc.Monitor().abortRequested:
	xbmc.sleep(1000)

del scan_monitor


default.py
# -*- coding: utf-8 -*-
import os, time
import json
import urllib, urllib2
import ssl
import xbmc
import xbmcaddon
import xbmcvfs

addon = xbmcaddon.Addon()
enable_debug = addon.getSetting('enable_debug')
if addon.getSetting('custom_path_enable') == "true" and addon.getSetting('custom_path') != "":
	txtpath = addon.getSetting('custom_path').decode("utf-8")
else:
	txtpath = xbmc.translatePath(addon.getAddonInfo('profile')).decode("utf-8")
	if not os.path.exists(txtpath):
		os.makedirs(txtpath)
txtfile = txtpath + "lastPlayed.json"
starmovies = addon.getSetting('starmovies.org')
enable_debug = addon.getSetting('enable_debug')
lang = addon.getLocalizedString

# Builds JSON request with provided json data
def buildRequest(method, params, jsonrpc='2.0', rid='1'):
	request = { 'jsonrpc' : jsonrpc, 'method' : method, 'params' : params, 'id' : rid }
	return request

# Checks JSON response and returns boolean result
def checkReponse(response):
	result = False
	if ( ('result' in response) and ('error' not in response) ):
		result = True
	return result

# Executes JSON request and returns the JSON response
def JSexecute(request):
	request_string = json.dumps(request)
	response = xbmc.executeJSONRPC(request_string)
	if ( response ):
		response = json.loads(response)
	return response

# Performs single JSON query and returns result boolean, data dictionary and error string
def JSquery(request):
	result = False
	data = {}
	error = ''
	if ( request ):
		response = JSexecute(request)
		if ( response ):
			result = checkReponse(response)
			if ( result ):
				data = response['result']
			else: error = response['error']
	return (result, data, error)

def send2starmovies(line):
	if enable_debug	== "true": xbmc.log("<<<plugin.video.last_played (starmovies) "+str(line), 3)
	wid = 0
	if line["id"]!="": wid = int(line["id"])
	if line["type"]=="movie": typ="M"
	elif line["type"]=="episode": typ="S"
	else: typ="V"

	if typ=="M" and addon.getSetting('movies') != "true": return
	if typ=="S" and addon.getSetting('tv') != "true": return
	if typ=="V" and addon.getSetting('videos') != "true": return

	imdbId = ""
	tvdbId = ""
	orgTitle = ""
	showTitle = line["show"]
	season = line["season"]
	episode = line["episode"]
	thumbnail = line["thumbnail"]
	fanart = line["fanart"]
	if wid>0:
		if typ=="M":
			request = buildRequest('VideoLibrary.GetMovieDetails', {'movieid' : wid, 'properties' : ['imdbnumber', 'originaltitle']})
			result, data = JSquery(request)[:2]
			if ( result and 'moviedetails' in data ):
				imdbId = data['moviedetails']["imdbnumber"]
				orgTitle = data['moviedetails']["originaltitle"]
		elif typ=="S":
			request = buildRequest('VideoLibrary.GetEpisodeDetails', {'episodeid' : wid, 'properties' : ['tvshowid', 'season', 'episode']})
			result, data = JSquery(request)[:2]
			if ( result and 'episodedetails' in data ):
				season = data['episodedetails']["season"]
				episode = data['episodedetails']["episode"]
				request = buildRequest('VideoLibrary.GetTvShowDetails', {'tvshowid' : data['episodedetails']["tvshowid"], 'properties' : ['imdbnumber', 'originaltitle']})
				result, data = JSquery(request)[:2]
				if ( result and 'tvshowdetails' in data ):
					showTitle = data['tvshowdetails']["label"]
					orgTitle = data['tvshowdetails']["originaltitle"]
					tvdbId = data['tvshowdetails']["imdbnumber"]

	xvideo = line["file"]
	if "video" in line and line["video"]!="": xvideo = line["video"]
	url = "https://www.starmovies.org/WebService.asmx/kodiWatch?tmdbId="
	url = url + "&tvdbId=" + tvdbId
	url = url + "&imdbId=" + imdbId
	url = url + "&kodiId=" + str(wid)
	url = url + "&title=" + urllib.quote(line["title"].encode("utf-8"))
	url = url + "&orgtitle=" + urllib.quote(orgTitle.encode("utf-8"))
	url = url + "&year=" + str(line["year"])
	url = url + "&source=" + urllib.quote(line["source"].encode("utf-8"))
	url = url + "&type=" + typ
	url = url + "&usr=" + urllib.quote(addon.getSetting('TMDBusr').encode("utf-8"))
	url = url + "&pwd=" + addon.getSetting('TMDBpwd')
	url = url + "&link=" + urllib.quote(xvideo.encode("utf-8"))
	url = url + "&thumbnail=" + urllib.quote(line["thumbnail"].encode("utf-8"))
	url = url + "&fanart=" + urllib.quote(line["fanart"].encode("utf-8"))
	url = url + "&showtitle=" + urllib.quote(showTitle.encode("utf-8"))
	url = url + "&season=" + str(season)
	url = url + "&episode=" + str(episode)
	url = url + "&version=1.19"
	url = url + "&date=" + line["date"]
	if enable_debug	== "true": xbmc.log("<<<plugin.video.last_played (starmovies) "+url, 3)
	try:
		context = ssl._create_unverified_context()
		request = urllib2.Request(url)
		response = urllib2.urlopen(request, context=context)
	except:
		pass

def videoEnd():
	retry=1
	xsource=''
	while xsource=='' and retry<50:
		xsource = xbmc.getInfoLabel('ListItem.Path').decode("utf-8")
		retry=retry+1
		time.sleep(0.1)

	if xsource=='': xsource="player"	
	xtitle = xbmc.getInfoLabel('ListItem.Title').decode("utf-8").strip()
	if xtitle=="" : xtitle = player_monitor.title
	xyear = xbmc.getInfoLabel('ListItem.Year')
	if xyear is None : xyear = player_monitor.year
	xid = xbmc.getInfoLabel('ListItem.DBID')
	if xid=="": xid = player_monitor.id
	xtype = xbmc.getInfoLabel('ListItem.DBTYPE')
	if xtype=="": xtype = player_monitor.type
	xthumb = xbmc.getInfoLabel('ListItem.Art(thumb)').decode("utf-8")
	if xthumb=="": xthumb = player_monitor.thumbnail
	xfanart = xbmc.getInfoLabel('ListItem.Art(fanart)').decode("utf-8")
	if xfanart=="": xfanart = player_monitor.fanart
	xshow = xbmc.getInfoLabel('ListItem.TVShowTitle').decode("utf-8").strip()
	xseason = xbmc.getInfoLabel('ListItem.Season')
	xepisode = xbmc.getInfoLabel('ListItem.Episode')
	try:
		if xshow=="": xshow = player_monitor.showtitle
		if xseason=="": xseason = player_monitor.season
		if xepisode=="": xepisode = player_monitor.episode
	except:
		pass
	xvideo = player_monitor.video		
	xfile = xbmc.getInfoLabel('ListItem.FileNameAndPath').decode("utf-8").strip()
	
	if xid!="" and int(xid)>0:
		if xtype=="movie": xsource=lang(30002)
		elif xtype=="episode": xsource=lang(30003)
		elif xtype=="musicvideo": xsource=lang(30004)
		else: xsource=xtype
	else:
		ads = xsource.split("/")
		if len(ads) > 2: xsource = ads[2]
	
	# if source is on blacklist, do not keep
	if addon.getSetting('blackadddon').lower().find(xsource.lower())>=0: return
	if addon.getSetting('blackfolder')!="":
		for dir in addon.getSetting('blackfolder').lower().split(","):
			if xsource.lower().find(dir)>=0: return
	if addon.getSetting('blackvideo')!="":
		for vid in addon.getSetting('blackvideo').lower().split(","):
			if xtitle.lower().find(vid)>=0: return
	
	if enable_debug	== "true": xbmc.log("<<<plugin.video.last_played (end source) "+xsource, 3)
	if xbmcvfs.exists(txtfile):
		f = xbmcvfs.File(txtfile)
		try: lines = json.load(f)
		except: lines = []
		f.close()
	else: lines = []

	replay = "N"
	for line in lines:
		if xfile!="" and xfile==line["file"]: replay = "S"
		if "video" in line and xvideo==line["video"]: replay = "S"
		if replay == "S":
			lines.remove(line)
			line.update({"date": time.strftime("%Y-%m-%d")})
			line.update({"time": time.strftime("%H:%M:%S")})
			lines.insert(0, line)
			replay = "S"
			if enable_debug	== "true": xbmc.log("<<<plugin.video.last_played (end final replay) "+str(line), 3)
			if starmovies	== "true": send2starmovies(line)
			break

	if replay=="N":
		newline = {"source":xsource, "title":xtitle, "year":xyear, "file":xfile, "video": xvideo, "id":xid, "type":xtype,"thumbnail":xthumb, "fanart":xfanart, "show":xshow, "season":xseason, "episode":xepisode, "date":time.strftime("%Y-%m-%d"), "time":time.strftime("%H:%M:%S")}
		lines.insert(0, newline)
		if enable_debug	== "true": xbmc.log("<<<plugin.video.last_played (end final play) "+str(newline), 3)
		if starmovies	== "true": send2starmovies(newline)
		if len(lines)>100:
			del lines[len(lines)-1]

	if len(lines)>0:
		f = xbmcvfs.File(txtfile, 'w')
		json.dump(lines, f)
		f.close()

class KodiPlayer(xbmc.Player):
	def __init__(self, *args, **kwargs):
		xbmc.Player.__init__(self)

	@classmethod
	def onPlayBackEnded(self):
		videoEnd()

	@classmethod
	def onPlayBackStopped(self):
		videoEnd()

	def onPlayBackStarted(self):
		player_monitor.video = player_monitor.getPlayingFile()
		request = {"jsonrpc": "2.0", "method": "Player.GetItem", "params": { "properties": ["title", "year", "thumbnail", "fanart", "showtitle", "season", "episode"], "playerid": 1 }, "id": "VideoGetItem"}
		result, data = JSquery(request)[:2]
		item=data["item"]
		if enable_debug	== "true": xbmc.log("<<<plugin.video.last_played (start play) "+str(item), 3)
		if "title" in item: player_monitor.title = item["title"]
		else: player_monitor.title = ""
		if player_monitor.title=="" and "label" in item: player_monitor.title = item["label"]		
		if "year" in item: player_monitor.year = item["year"]
		else: player_monitor.year = ""
		if "thumbnail" in item: player_monitor.thumbnail = item["thumbnail"]
		else: player_monitor.thumbnail = ""
		if "fanart" in item: player_monitor.fanart = item["fanart"]
		else: player_monitor.fanart = ""
		if "showtitle" in item: player_monitor.showtitle = item["showtitle"]
		else: player_monitor.showtitle = ""
		if "season" in item and item["season"]>0: player_monitor.season = item["season"]
		else: player_monitor.season = ""
		if "episode" in item and item["episode"]>0: player_monitor.episode = item["episode"]
		else: player_monitor.episode = ""
		if "id" in item: player_monitor.id = item["id"]
		else: player_monitor.id = ""
		if "type" in item: player_monitor.type = item["type"]
		else: player_monitor.type = ""
		
player_monitor = KodiPlayer()

while not xbmc.abortRequested:
	xbmc.sleep(1000)

del player_monitor
